pkg/containertools/dockerfilegenerator.go

```diff
@@ -0,0 +1,98 @@
+package containertools
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	baseImage = "scratch"
+	defaultBinarySourceImage = "quay.io/kevinrizza/upstream-opm-builder"
+	defaultDbLocation = "./index.db"
+	dbLocationLabel = "operators.operatorframework.io.index.database.v1"
+)
+
+// DockerfileGenerator defines functions to generate index dockerfiles
+type DockerfileGenerator interface {
+	GenerateIndexDockerfile(string,string,[]string, []string) (string, error)
+}
+
+// DockerfileGeneratorImpl struct implementation of DockerfileGenerator interface
+type DockerfileGeneratorImpl struct {
+	logger *logrus.Entry
+	labelReader LabelReader
+}
+
+// NewDockerfileGenerator is a constructor that returns a DockerfileGenerator
+func NewDockerfileGenerator(containerTool string, logger *logrus.Entry) DockerfileGenerator {
+	return &DockerfileGeneratorImpl{
+		logger: logger,
+		labelReader: NewLabelReader(containerTool, logger),
+	}
+}
+
+// GenerateIndexDockerfile builds a string representation of a dockerfile to use when building
+// an operator-registry index image
+func (g *DockerfileGeneratorImpl) GenerateIndexDockerfile(binarySourceImage, previousIndexImage string, bundlesToAdd, operatorsToDelete []string) (string, error) {
+	var dockerfile string
+
+	if binarySourceImage == "" {
+		binarySourceImage = defaultBinarySourceImage
+	}
+
+	// Builder Image(s)
+	// If building from a previous database, use it as a builder to copy old registry database
+	if previousIndexImage != "" {
+		dockerfile += fmt.Sprintf("FROM %s as previous-index\n\n", previousIndexImage)
+
+		// BUILDER Binary source image that defines where to get `opm` from
+		dockerfile += fmt.Sprintf("FROM %s AS builder\n", binarySourceImage)
+
+		previousDbLocation := defaultDbLocation
+
+		g.logger.Infof("Pulling previous image %s to get metadata", previousIndexImage)
+
+		// Get the old index image's dbLocationLabel to find this path
+		labels, err := g.labelReader.GetLabelsFromContainer(previousIndexImage)
+		if err != nil {
+			return "", err
+		}
+		if dbLocation, ok := labels[dbLocationLabel]; ok {
+			previousDbLocation = dbLocation
+		}
+
+		// copy the old database file in the scratch image
+		dockerfile += fmt.Sprintf("COPY --from=previous-index %s /previous-index.db\n", previousDbLocation)
+	} else { // Otherwise just define the builder as a place to get the binary and generate the database
+		dockerfile += fmt.Sprintf("FROM %s AS builder\n", binarySourceImage)
+	}
+
+	// Generate the registry database
+	if len(bundlesToAdd) > 0 {
+		add := strings.Join(bundlesToAdd, ",")
+		dockerfile += fmt.Sprintf("RUN /build/bin/opm registry add \"%s\" -o \"index.db\"\n", add)
```

Comment:
nit: seems weird to intentionally place our binaries in `/build`?

Reply:
Yeah this is an artifact of the fact that the builder image (https://github.com/operator-framework/operator-registry/blob/master/upstream-builder.Dockerfile) just runs `make` on the source. If we wanted to change that we may have to create a new dockerfile for that build since I assume other things depend on that one.

Reply:
Ah, I see that in the final image it's in a better location. Thanks!